bpf/bpf_wireguard.c

```diff
@@ -12,16 +12,368 @@
+					   METRIC_INGRESS, false, false, 0);
+	}
+
+	return TC_ACT_OK;
+}
+
+static __always_inline int
+handle_ipv4(struct __ctx_buff *ctx, __u32 identity __maybe_unused,
+	    bool *punt_to_stack __maybe_unused, __s8 *ext_err __maybe_unused)
+{
+	void *data_end, *data;
+	struct iphdr *ip4;
+
+	if (!revalidate_data_pull(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+/* If IPv4 fragmentation is disabled
+ * AND a IPv4 fragmented packet is received,
+ * then drop the packet.
+ */
+#ifndef ENABLE_IPV4_FRAGMENTS
+	if (ipv4_is_fragment(ip4))
+		return DROP_FRAG_NOSUPPORT;
+#endif
+
+#ifdef ENABLE_NODEPORT
+	if (!ctx_skip_nodeport(ctx)) {
+		bool is_dsr = false;
+		int ret = nodeport_lb4(ctx, ip4, ETH_HLEN, identity, punt_to_stack,
+				   ext_err, &is_dsr);
+#ifdef ENABLE_IPV6
+			if (ret == NAT_46X64_RECIRC) {
+				ctx_store_meta(ctx, CB_SRC_LABEL, identity);
+				return tail_call_internal(ctx, CILIUM_CALL_IPV6_FROM_WIREGUARD,
+							  ext_err);
+			}
+#endif
+		/* nodeport_lb4() returns with TC_ACT_REDIRECT for
+		 * traffic to L7 LB. Policy enforcement needs to take
+		 * place after L7 LB has processed the packet, so we
+		 * return to stack immediately here with
+		 * TC_ACT_REDIRECT.
+		 */
+		if (ret < 0 || ret == TC_ACT_REDIRECT)
+			return ret;
+		if (*punt_to_stack)
+			return ret;
+	}
+#endif
+
+	return TC_ACT_OK;
+}
+
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV4_FROM_WIREGUARD)
+int tail_handle_ipv4(struct __ctx_buff *ctx)
+{
+	__u32 src_sec_identity = ctx_load_and_clear_meta(ctx, CB_SRC_LABEL);
+	__s8 ext_err = 0;
+	bool punt_to_stack = false;
+	int ret;
+
+	ret = handle_ipv4(ctx, src_sec_identity, &punt_to_stack, &ext_err);
+	if (ret == CTX_ACT_OK) {
+		if (punt_to_stack)
+			return ret;
+		ret = handle_ipv4_cont(ctx, src_sec_identity, &ext_err);
+	}
+
+	if (IS_ERR(ret))
+		return send_drop_notify_error_ext(ctx, src_sec_identity, ret, ext_err,
+						METRIC_INGRESS);
+	return ret;
+}
+#endif /* ENABLE_IPV4 */
+
+/* from-wireguard is attached as a tc ingress filter to the cilium_wg0 device.
+ */
+__section_entry
+int cil_from_wireguard(struct __ctx_buff *ctx)
+{
+	void __maybe_unused *data, *data_end;
+	struct ipv6hdr __maybe_unused *ip6;
+	struct iphdr __maybe_unused *ip4;
+
+	int ret = TC_ACT_OK;
+	__u32 identity = UNKNOWN_ID;
+	__u16 proto = 0;
+	__s8 ext_err = 0;
+
+	if (!validate_ethertype(ctx, &proto)) {
+		/* Pass unknown traffic to the stack */
+		goto out;
+	}
+
+	switch (proto) {
+#ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		if (!revalidate_data_pull(ctx, &data, &data_end, &ip6))
+			return send_drop_notify_error(ctx, identity, DROP_INVALID, METRIC_INGRESS);
+
+		identity = resolve_srcid_ipv6(ctx, ip6);
+		send_trace_notify(ctx, TRACE_FROM_CRYPTO, identity, UNKNOWN_ID,
+				  TRACE_EP_ID_UNKNOWN, ctx->ingress_ifindex,
+				  TRACE_REASON_UNKNOWN, TRACE_PAYLOAD_LEN);
+
+		ret = tail_call_internal(ctx, CILIUM_CALL_IPV6_FROM_WIREGUARD, &ext_err);
```

Comment:
I think here we can directly call `tail_handle_ipv6`: this PR removes a huge portion of the codebase from `cil_from_netdev` (e.g., Host Firewall), therefore shouldn't hit complexity issue.